### ISSUE TITLE
Filename comparisons when finding fcb are always case-sensitive

### ISSUE THREAD

**Issue by bailey27** [dokan-dev/dokany#294]
# Environment
 * Windows version:  Windows 10
 * Processor architecture: x64
 * Dokany version: current source
 * Library type (Dokany/FUSE):  Dokany
 
# Check List
  - [x ] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x ] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x ] I can always reproduce the issue with the provided description below.
  - [ x] I have updated Dokany to the latest version (https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.

# Description

The issue is that, in the kernel-mode driver, when DokanGetFCB() goes to find an existing fcb for a file, it does the string comparison in a case-sensitive fashion.

I think this generally isn't a problem, but, because NTFS is (normally) case-insensitive, you could end up with two FCBs for the same file, which isn't how it's supposed to be.

I was able to get it to happen
```
[DokanFS]   Create: ProcessId 612, FileName:\FOO.txt
[DokanFS]   DokanGetFCB has entry FileName: \ FileCount: 3. Looking for \FOO.txt
[DokanFS]   DokanGetFCB has entry FileName: \foo.txt FileCount: 1. Looking for \FOO.txt
[DokanFS]   Allocate FCB for \FOO.txt
[DokanFS]   Create: FileName:\FOO.txt, status = 0x00000103
```

As you can see, even though there is already an fcb for \\foo.txt, it is allocating a new one for \\FOO.txt.

I changed the code to use RtlEqualUnicodeString() and also use the SL_CASE_SENSITIVE in irpSp->Flags to determine if the comparison should be case-sensitive.

Here is my branch
https://github.com/bailey27/dokany/commit/5960d59a1fa0d1f09d61bc627e8a805cd10c8586

Now it finds the fcb for \\foo.txt when it goes to open \\FOO.txt

```
[DokanFS]   Create: ProcessId 4736, FileName:FOO.txt
[DokanFS]   DokanGetFCB has entry FileName: \ FileCount: 2. Looking for \FOO.txt
[DokanFS]   DokanGetFCB has entry FileName: \foo.txt FileCount: 1. Looking for \FOO.txt
[DokanFS]   Found existing FCB for \FOO.txt
[DokanFS]   Create: FileName:FOO.txt, status = 0xc0000043

```

As you can see, the driver now detects the sharing violation.  Before, it was sending the create to user mode, and the sharing violation was being detected there.

# Logs
see above

**Comment by Liryna** [comment]
Hi @bailey27 ,

Well found and thank you for your contribution as always ! :+1:

You are right, `SL_CASE_SENSITIVE` has to be handle in the way that you propose it !

You can create a pull request :)

**Comment by bailey27** [comment]
@Liryna ,

Thanks.  I made the pull request.

**Comment by bailey27** [comment]
@Liryna ,

I think I see what you meant before about my commit history.

I think I need to rebase with upstream instead of merge.

I'll try that and see if it cleans up the commit history, and and if so submit a new pull request.

**Comment by Liryna** [comment]
:smiley: I was going to say it

**Comment by bailey27** [comment]
@Liryna ,

I think I reset the commit history ok now.  Please tell me if I did it wrong.  I need to learn how to do this.

I made a new pull request.

**Comment by Liryna** [comment]
Done with your pull request https://github.com/dokan-dev/dokany/pull/297

Thank you @bailey27 for the report and the patch at the same time ! :heart: very appreciate !

**Comment by Liryna** [comment]
Hey @bailey27

I made new changes to clearly define if the drive is case sensitive or insensitive but for case insensitive I still would like to support the sl case sensitive flag / file posix case sensitive flag.

I am having issue to pass those flags to the dokan driver like if they were removed by the io manager before reaching the driver. Do you remember how you succeeded to pass those flags?

Thank in advance!

**Comment by Liryna** [comment]
@bailey27 OK found!
Need to set HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\kernel\ dword:ObCaseInsensitive registry value to 0 and only after the flag is passed to the driver.

@DDoSolitary just seen you here https://github.com/microsoft/WSL/issues/2954 Do you have any idea if that really interesting to support SL_CASE_SENSITIVE / FILE_FLAG_POSIX_SEMANTICS for dokan for the few system who would dare changing ObCaseInsensitive ?

**Comment by DDoSolitary** [comment]
@Liryna I haven't tested anything yet, but I guess support for these flags is required if someone wants to install WSL on a Dokan drive. I haven't seen any other Windows components or third-party softwares that makes use of these flags, though.

**Comment by Liryna** [comment]
@DDoSolitary WSL is really using this flag an not set the new case sensitive information perf directory?

**Comment by DDoSolitary** [comment]
@Liryna Oh I forgot about the fact that the new per-directory flag works even without SL_CASE_SENSITIVE. Then these flags are no longer required for WSL in recent versions of Windows 10 (I'm not very sure about the exact version, probably since ver 1803)

**Comment by bailey27** [comment]
@Liryna I don't have anything to contribute now.  I just want to say Hi! and thanks for keeping Dokany alive and well.
